File: backend/internal/service/trip_service.go
package service

import (
	"github.com/ecotrip/backend/internal/model"
	"github.com/ecotrip/backend/internal/repository"
)

type TripService interface {
	GetTrips(userID uint) ([]model.Trip, error)
	GetTrip(id, userID uint) (*model.Trip, error)
	CreateTrip(userID uint, title, description, destination, startDate, endDate string) (*model.Trip, error)
	DeleteTrip(id, userID uint) error
}

type tripService struct {
	tripRepo repository.TripRepository
}

func NewTripService(tripRepo repository.TripRepository) TripService {
	return &tripService{tripRepo: tripRepo}
}

func (s *tripService) GetTrips(userID uint) ([]model.Trip, error) {
	return s.tripRepo.FindAll(userID)
}

func (s *tripService) GetTrip(id, userID uint) (*model.Trip, error) {
	return s.tripRepo.FindByID(id, userID)
}

func (s *tripService) CreateTrip(userID uint, title, description, destination, startDate, endDate string) (*model.Trip, error) {
	trip := &model.Trip{
		UserID:      userID,
		Title:       title,
		Description: description,
		Destination: destination,
		StartDate:   startDate,
		EndDate:     endDate,
	}
	if err := s.tripRepo.Create(trip); err != nil {
		return nil, err
	}
	return trip, nil
}

func (s *tripService) DeleteTrip(id, userID uint) error {
	return s.tripRepo.Delete(id, userID)
}

File: frontend/services/tripService.ts
import api from './api';
import { DUMMY_MODE } from '@/constants/config';

export interface Trip {
  id: number;
  title: string;
  description: string;
  destination: string;
  start_date: string;
  end_date: string;
  carbon_footprint: number;
  created_at: string;
}

export interface CreateTripPayload {
  title: string;
  description: string;
  destination: string;
  start_date: string;
  end_date: string;
}

export const tripService = {
  async getTrips(): Promise<Trip[]> {
    if (DUMMY_MODE) return [];
    const res = await api.get<Trip[]>('/trips');
    return res.data;
  },

  async getTrip(id: number): Promise<Trip> {
    if (DUMMY_MODE) {
      return {
        id,
        title: 'Dummy Trip',
        description: '',
        destination: 'Dummy Destination',
        start_date: '2026-01-01',
        end_date: '2026-01-02',
        carbon_footprint: 0,
        created_at: '2026-01-01',
      };
    }
    const res = await api.get<Trip>(`/trips/${id}`);
    return res.data;
  },

  async createTrip(payload: CreateTripPayload): Promise<Trip> {
    if (DUMMY_MODE) {
      return {
        id: Date.now(),
        ...payload,
        carbon_footprint: 0,
        created_at: new Date().toISOString(),
      };
    }
    const res = await api.post<Trip>('/trips', payload);
    return res.data;
  },

  async deleteTrip(id: number): Promise<void> {
    if (DUMMY_MODE) return;
    await api.delete(`/trips/${id}`);
  },
};
